README.md
![logo](https://github.com/YeoLab/rna_editing_pipeline/blob/master/logo/sailor-logo.png)

# Software for Accurately Identifying Locations Of RNA-editing (SAILOR)

SAILOR implements published methodologies to assess adenosine to inosine changes in RNA-SEQ data for easy identification of transcriptome-wide editing. 
The SAILOR pipeline is available as both a [CWL workflow](http://www.commonwl.org) and as a [Singularity container](http://singularity.lbl.gov) and is designed for ease of use to run with one command. 
It requires a [BAM](https://genome.ucsc.edu/FAQ/FAQformat.html#format5.1)-formatted file of the sequence alignments, a [FASTA](http://genetics.bwh.harvard.edu/pph/FASTA.html)-formatted reference genome sequence (of any organism or cell-type), 
and a [BED3](https://genome.ucsc.edu/FAQ/FAQformat.html#format1)-formatted file of known SNPs. SAILOR allows the user to specify a range of filtering criteria including: Non A-to-I mismatch rate, 
location of mismatches (to account for biases at the end of reads), and a minimum read coverage required to call variants. Users may relax any 
of these filtering criteria and/or pursue analysis of A-to-I editing sites with lower confidence scores.

# Installation:

[Install singularity](http://singularity.lbl.gov/install-linux)

[Download executable into an empty directory](https://s3-us-west-1.amazonaws.com/sailor-1.0.4/sailor-1.0.4)

That's it!

# Running the Pipeline:

Mark downloaded programme as executable:
```
chmod +x sailor-1.0.4
```

Display  a brief overview of options and expected parameters:
```
./sailor-1.0.4
```
Type 'y' to populate the current folder with example files:

```
Do you want me to copy now into your current directory an example of data input files and a job template (y/-):y
```

Execute the analysis using the provided example YAML file pointing to the appropriate bundled example files
```
./sailor-1.0.4 ce11_example.yaml
```

### Running the data with required arguments:

Running time for the examples should be quick! Running on a complete dataset takes a few hours for C elegans data, 
so sit back and relax by reading the rest of this README.

These are the minimum required arguments needed to run the pipeline (you can view the same information inside the example.yml file):


This is a BAM file ([Example](https://github.com/YeoLab/sailor/blob/master/CWL-SINGULARITY-pipeline-building-code/example/ce11_example_single_end.bam)) of your reads aligned to the genome. You can generate this file using any short read aligner, and it does not need to be sorted (the pipeline will split + sort things for you). **Note: You should check whether your aligner provides the MD tag, as this is necessary to determine the non A/G mismatches between the reference and query.** Our example.bam is a downsampled BAM file containing the first 10,000 lines (9,983 reads) of a real sample:
```YAML
input_bam:
  class: File
  path: ce11_example_single_end.bam
```


This describes the reference genome in FASTA format ([Example](https://github.com/YeoLab/sailor/blob/master/CWL-SINGULARITY-pipeline-building-code/example/ce11.chrI.fa)) (used in the mpileup step), which specifies the reference for which variant reads are compared against. The included reference is the first chromosome of a ce11 assembly:
```YAML
reference:
  class: File
  path: ce11.chrI.fa
```


This file contains a list of known SNPs ([Example](https://github.com/YeoLab/sailor/blob/master/CWL-SINGULARITY-pipeline-building-code/example/ce11_known_SNPs.bed)) which will be filtered from the list of candidate editing sites. The example file contains just one SNP in BED3 format (0-based, half-open), which can be used to remove sites that we know aren't editing sites, but are known SNPs:
```YAML
known_snp:
  class: File
  path: ce11_known_SNPs.bed
```


### Running the data with optional arguments:

The pipeline only requires 3 arguments: The BAM file, the reference genome, and a list of known SNPs to filter out. 
If you find that using default parameters are not fit to your data, you may want to play around with the optional arguments. 
Here is the full list and explanation of optional arguments you can provide as needed in the job file 
(You can find an example YAML configuaration file [here](https://github.com/YeoLab/sailor/blob/master/CWL-SINGULARITY-pipeline-building-code/example/ce11_example.yaml)):

This parameter [true] or false specifies whether or not we're dealing with a reversely stranded library:
```YAML
reverse_stranded_library: true
```

This option [true] or false specifies whether or not the reads are single or paired end. This is equivalent to the [-s option](http://www.htslib.org/doc/samtools.html) of ```samtools rmdup``` step, which is part of the workflow:
```YAML
single_end: true
```

This specifies how much overlap is minimally required for a read spanning a junction to be counted. What this really means is when aligned to a reference, if there is a gap between the read (aka spans an intron), we want to ensure there is sufficient overlap in the two sides spanning that gap. See [CIGAR 'N'](https://samtools.github.io/hts-specs/SAMv1.pdf) for more details:
```YAML
junction_overhang: 10
```

This removes reads with mutations this many nucleotides away from the read. We want to avoid any variations at the read ends where sequencing errors and technical artifacts are more likely to occur:
```YAML
edge_mutation: 5
```

This specifies the maximum amount of non-AG variations (or TC antisense) we expect per read. If we see multiple AG variants, great! However if we see many unexpected non-AG variants on a per-read basis, we can conservatively remove them from downstream analysis:
```YAML
# remove reads with more than [1] unexpected mutation (one that isn't A-G or T-C antisense).
non_ag: 1
```

This provides a hard filter for variants prior to scoring. By default, for any site to be considered, it must have a total coverage of 5. In this workflow, total coverage is calculated by either the DP flag (specifying coverage), or the 'DP4' flag, specifying "high quality" coverage. This option can be set with the dp: field. See samtools [mpileup](http://www.htslib.org/doc/samtools.html) for more info:
```YAML
min_variant_coverage: 5
dp: DP4
```

These parameters can relax or tighten the beta distribution curve used to score confident editing sites. Essentially these numbers adjust the pseudocount added to variant site coverage. Therefore increasing the alpha and beta parameters will generally relax coverage requirements, which is sometimes useful for low-coverage data. See the [original paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3959997/) for more info:
```YAML
alpha: 0
beta: 0
edit_fraction: 0.01
```

This parameter specifies options to either 1) keep 100% edited sites as confident 'edited' sites (true), or 2) flag as 'possible snps' (false):
```YAML
keep_all_edited: false
```

This parameter specifies whether or not you want to skip the 'samtools rmdup' step of the pipeline:
```YAML
skip_duplicate_removal: false
```

# Outputs:
There are lots of intermediate files, but the ones we want are the bed files (it's a long name, but it helps trace all the steps and filters that the pipeline goes through):
```
SAMPLENAME.fwd.sorted.rmdup.readfiltered.formatted.varfiltered.snpfiltered.ranked.bed
SAMPLENAME.rev.sorted.rmdup.readfiltered.formatted.varfiltered.snpfiltered.ranked.bed
```
These BED6 files correspond to candidate editing sites found at either the positive strand (fwd) or negative strand (rev):

### Format of the BED file:

1. chromosome
2. start (0-based) index of an editing site
3. end (open) index of an editing site
4. unique name containing information about coverage|variant type|edit% (```84|A>G|0.011904762``` corresponds to an A>G (+) site covered by 84 reads that is ~1% edited)
5. confidence score
6. strand

You can load these BED files onto [IGV](http://software.broadinstitute.org/software/igv/download) and see if your gene of interest is edited.
For a more global viewpoint, you will need to filter this BED file based on the 'confidence score' (tab 5) using anything that can sort/filter a tabbed file. Then you can look at these filtered sites on IGV.

If you don't see many sites, you can relax the parameters, or look into the intermediate files described below:

If our sample is ```sample.bam```, we expect to obtain a list of the following outputs (for both fwd and rev stranded edits):


sorted BAM file:

```
example.fwd.sorted.bam
```


sorted BAM file with 'duplicate reads' removed (defined as reads sharing the same external mapped coordinates, keeping reads with the highest mapping quality (see [samtools](http://www.htslib.org/doc/samtools.html)):

```
example.fwd.sorted.rmdup.bam
```


The above BAM file filtered of reads that do not pass read-centric thresholds:

```
example.fwd.sorted.rmdup.readfiltered.bam
```


The above BAM file as pileup in gbcf (binary) format:

```
example.fwd.sorted.rmdup.readfiltered.gbcf
```


The above gbcf file in human-readable vcf format:

```
example.fwd.sorted.rmdup.readfiltered.vcf
```


The above vcf file in a more familiar vcf format:

```
example.fwd.sorted.rmdup.readfiltered.formatted.vcf
```


The above vcf file filtered of SNVs that do not pass the position-centric thresholds:

```
example.fwd.sorted.rmdup.readfiltered.formatted.varfiltered.vcf
```


The above vcf file filtered of SNVs that are also known SNPs:

```
example.fwd.sorted.rmdup.readfiltered.formatted.varfiltered.snpfiltered.vcf
```


The above vcf file in a tabular 'confidence' format, showing:

```
example.fwd.sorted.rmdup.readfiltered.formatted.varfiltered.snpfiltered.ranked.conf
```

##### Format (tabs) of the conf file:
1. (#CHROM) : chromosome
2. (POS) : 1-based position of the editing site
3. (NUM_READS) : number of total coverage
4. (REF) : reference allele
5. (ALT) : alternate allele
6. (CONFIDENCE) : confidence score
7. (POST_PSEUDOCOUNT_EDIT%) : if we add pseudocounts, the edit % will be here
8. (PRE_PSEUDOCOUNT_EDIT%) : native edit %
9. (FILTER) : either PASS for a valid editing candidate (regardless of score), or SNP if the site was 100% A>G
10. (INFO) : vcf "info" column (see [vcf](https://samtools.github.io/hts-specs/VCFv4.2.pdf) format for details)
11. (GENOTYPE) : vcf "genotype" column
12. (baz) : vcf "genotype value" column


## Other notes.

If using python3.8, you might run into a warning related to buffer size (ie. 

```bash
subprocess.py:849: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used
```

)

This was identified here and doesn't appear to affect this pipeline.

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=954388

##

# References:

Washburn, M. C., Kakaradov, B., Sundararaman, B., Wheeler, E., Hoon, S., Yeo, G. W., & Hundley, H. A. (2014): The dsRBP and inactive editor ADR-1 utilizes dsRNA binding to regulate A-to-I RNA editing across the C. elegans transcriptome. 
Cell reports, 6(4), 599-607.

Amstutz, Peter; Crusoe, Michael R.; Tijanić, Nebojša; Chapman, Brad; Chilton, John; Heuer, Michael; Kartashov, Andrey; Leehr, Dan; Ménager, Hervé; Nedeljkovich, Maya; Scales, Matt; Soiland-Reyes, Stian; Stojanovic, Luka (2016): Common Workflow Language, v1.0. 
figshare. https://doi.org/10.6084/m9.figshare.3115156.v2
Retrieved: 22 13, May 11, 2017 (GMT)

Kurtzer GM, Sochat V, Bauer MW (2017): Singularity: Scientific containers for mobility of compute. 
PLoS ONE 12(5): e0177459. https://doi.org/10.1371/journal.pone.0177459



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.99.0] - 10-21-2020
### Added
- post-processing steps for STAMP analysis:
    - combine_and_reformat
    - annotate
    - bam_to_bedgraph
    - bedgraphtobigwig
    - score_edits_total_coverage

## [1.1.0] - 11-05-2019
### Added
- added C>T editing site discovery (adds a ct flag to the main manifest)

## [1.0.4] - 5-12-2017
### Added
- added ...

### Changed
- changed ...

### Removed
- removed ...

### Fixed
- fixed paths in README

## [0.0.3DEV] - 5-4-2017
### Added
- BED file outputs to replace conf file as the final output

### Changed
- changed the 'reverse-strand' flag in filter_variants.py to '--reverse-split'. --reverse-split is true if the pipeline deals with '*.rev.bam' files generated from step1.

### Removed
- removed reverse-strand flags for filter_variants.py : we now get strand information based on how the mapped reads were split in the beginning of the workflow.

### Fixed
- changing the reverse flag in filter_variants now fixes instances where A-G changes were being called falsely on Paired End reads mapped to the reverse gene.


## [0.0.3DEV] - 5-9-2017
### Added
- example yaml files for PE and SE
- keep-100-edited flag to the 'rank_edits.py' script to keep or discard 100% edited SNVs

### Changed
- changed ...

### Removed
- removed ...

### Fixed
- fixed paths in README

## [0.0.3DEV] - 5-4-2017
### Added
- BED file outputs to replace conf file as the final output

### Changed
- changed the 'reverse-strand' flag in filter_variants.py to '--reverse-split'. --reverse-split is true if the pipeline deals with '*.rev.bam' files generated from step1.

### Removed
- removed reverse-strand flags for filter_variants.py : we now get strand information based on how the mapped reads were split in the beginning of the workflow.

### Fixed
- changing the reverse flag in filter_variants now fixes instances where A-G changes were being called falsely on Paired End reads mapped to the reverse gene.

## [0.0.3DEV] - 5-1-2017
### Added
- reverse-strand flags for:
  - filter_reads.py
  - filter_variants.py
  - split_strands.py
- new test_cases for filter_reads
- new test_cases for filter_variants
- new test_cases for split_strands
- example runscripts and job files to 0.0.3DEV versions for PE

### Changed
- cwl tools now reflect and have the option for reverse strand libraries (untested for forward)
- moved complete single-end example example_outputs folder
- added PE example data
- changed 'fwd' in reverse strand libraries to output A-G editing sites, and 'rev' in reverse strand libraries to output T-C (previously the opposite).

### Removed
- removed ...

### Fixed
- get_junction_overhangs() now checks for multiple junctions per read (ie. 34M10N34M10N34M)

## [0.0.3DEV] - 2017-04-27
### Added
- example runscripts and job files 0.0.3DEV versions
- example input and output files to 0.0.2

### Changed
- moved cwl-1.0 and cwl-draft3 to 'singularity-0.0.1'
- added a better README file

### Removed
- removed ...

### Fixed
- fixed ...

## [0.0.2] - 2017-03-04
### Added
- example runscripts and job files 0.0.2 versions
- example input and output files to 0.0.2

### Changed
- changed ...

### Removed
- removed ...

### Fixed
- fixed ...


pyscripts/rank_edits.py
#!/usr/bin/env python
# encoding: utf-8
'''
editing.rank_edits -- BAM filtering script

Given a VCF file, rank potential editing candidates given percentage of
editing and coverage supporting the site.

@author:     boyko

@copyright:  2016 yeolab. All rights reserved.

@license:    license

@contact:    bay001@ucsd.edu
@deffield    updated: 4-21-2017
'''

import sys
import os
import re
from scipy.special import betainc

from argparse import ArgumentParser
from argparse import RawDescriptionHelpFormatter

__all__ = []
__version__ = 0.1
__date__ = '2016-07-13'
__updated__ = '2016-07-13'


class CLIError(Exception):
    """
    Generic exception to raise and log different fatal errors.
    """

    def __init__(self, msg):
        super(CLIError).__init__(type(self))
        self.msg = "E: %s" % msg

    def __str__(self):
        return self.msg

    def __unicode__(self):
        return self.msg


def as_bed(line):
    """
    Returns the BED6-formatted line representation of the CONF file.
    :param line:
    :return:
    """
    line = line.split('\t')
    if line[3] == 'T' and line[4] == 'C':
        strand = '-'
    elif line[3] == 'A' and line[4] == 'G':
        strand = '+'
    else:
        print('WARNING: not a very good AG editing candidate!')
        strand = '0'
    bed_name = '{}|{}>{}|{}'.format(
        line[2], line[3], line[4], line[6]
    )
    return '{}\t{}\t{}\t{}\t{}\t{}\n'.format(
        line[0], int(line[1])-1, int(line[1]), bed_name, line[5], strand
    )


def as_vcf(line):
    """
    Returns a VCF-formatted line representation of the CONF file
    :param line:
    :return:
    """
    line = line.split('\t')
    var_identifier = 'cov_{}|editID_{}:{}'.format(line[2], line[0], line[1])
    return '{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}'.format(
        line[0], line[1], var_identifier, line[3], line[4], line[5],
        line[8], line[9], line[10], line[11]
    )


def process(alfa, beta, cov_margin, keep_all_edited, line):
    """
    Calculates, for a single line in a VCF formatted file, the
    confidence score based on depth of coverage and edit fraction %.

    :param line: string
        single vcf formatted line.
    :return confidence: float
        confidence value of the line
    :return return_string: basestring
        full vcf formatted line with confidence
    """
    (chr, pos, dot, ref, alt, qual,
     filter, info, format, cond) = line.split("\t")[:10]

    if chr[0] == "#":
        print line,
        return

    # retrieve total number of reads mapping to position
    infos = info.split(";")
    (dp, i16) = infos[:2]

    assert dp[:2] == "DP"
    num_reads = int(dp[3:])

    """
    # retrieve numbers of A's and G's on forward and reverse strand
    assert i16[:3] == "I16", i16
    (a_fwd, a_rev, g_fwd, g_rev) = (int(x) for x in i16[4:].split(",")[:4])
    print("warning: i16 not available")
    """

    dp4 = re.findall("DP4\=([\d\,]+)", info)[0]

    (a_fwd, a_rev, g_fwd, g_rev) = (int(x) for x in dp4.split(","))

    a = a_fwd + a_rev
    g = g_fwd + g_rev
    num_reads = a + g
    edit_frac = g / float(num_reads)

    # calc smoothed counts and confidence
    G = g + alfa
    A = a + beta
    theta = G / float(G + A)

    ########  MOST IMPORTANT LINE  ########
    # calculates the confidence of theta as
    # P( theta < cov_margin | A, G) ~ Beta_theta(G, A)
    confidence = 1 - betainc(G, A, cov_margin)

    # keep 100% edited sites or toss
    if A == 0 and not keep_all_edited:
        confidence = 0
        region = 'SNP'
    else:
        region = 'PASS'

    # print line in CONF format

    return_string = ("\t".join([chr, pos, str(num_reads), ref, alt, ""]) +
                     "\t".join(str(round(y, 9)) for y in [
                         confidence, theta, edit_frac
                     ]) +
                     "\t".join(["", region, info, format, cond]) +
                     "\n")
    return return_string


def rank_edits(alfa, beta, cov_margin, keep_all_edited, eff_file, outfile):
    """
    Step 8: Score editing sites based on coverage and edit %

    * Boyko's confidence scoring using an inverse probability model.

    :param alfa: int
        add A pseudocount
    :param beta: int
        add G pseudocount
    :param cov_margin: float
        minimum allowable edit fraction
    :param keep_all_edited: boolean
        report 100% edited sites
    :param eff_file:
        input vcf or snpEFF (vcf format) file
    :param outfile:
        output conf (vcf format) file
    :return:

    """
    print("Ranking Editing Sites: {}".format(eff_file))

    o = open(outfile, 'w')
    ob = open(os.path.splitext(outfile)[0] + '.bed', 'w')
    # ov = open(os.path.splitext(outfile)[0] + '.vcf', 'w')
    with open(eff_file, 'r') as f:
        for eff in f:
            if eff.startswith("\#\#") or eff.startswith("##"):
                pass
                # ov.write(eff)
            elif eff.startswith("#CHROM"):  # ammend header of EFF file
                # TODO: implement VCF file
                line = eff.split("\t")  # to contain extra numeric columns
                line[2] = "NUM_READS"
                line[5] = "PRE_PSEUDOCOUNT_EDIT%"
                line.insert(5, "POST_PSEUDOCOUNT_EDIT%")
                line.insert(5, "CONFIDENCE")
                o.write('\t'.join(line))
                # ov.write(eff)
            else:
                to_string = process(
                    alfa, beta, cov_margin, keep_all_edited, eff
                )
                ob.write(as_bed(to_string))
                # ov.write(as_vcf(to_string))
                o.write(to_string)
    o.close()
    ob.close()
    # ov.close()


def main(argv=None):  # IGNORE:C0111
    """
    Command line options.
    :param argv:
    :return:
    """

    if argv is None:
        argv = sys.argv
    else:
        sys.argv.extend(argv)

    program_name = os.path.basename(sys.argv[0])
    program_version = "v%s" % __version__
    program_build_date = str(__updated__)
    program_version_message = '%%(prog)s %s (%s)' % (
        program_version, program_build_date
    )
    program_shortdesc = __import__('__main__').__doc__.split("\n")[1]
    program_license = '''%s

  Created by user_name on %s.
  Copyright 2016 organization_name. All rights reserved.

  Licensed under the Apache License 2.0
  http://www.apache.org/licenses/LICENSE-2.0

  Distributed on an "AS IS" basis without warranties
  or conditions of any kind, either express or implied.

USAGE
''' % (program_shortdesc, str(__date__))

    # Setup argument parser
    parser = ArgumentParser(
        description=program_license,
        formatter_class=RawDescriptionHelpFormatter
    )
    parser.add_argument(
        '-V', '--version',
        action='version',
        version=program_version_message
    )
    parser.add_argument(
        "-i", "--input",
        dest="input_eff",
        help="variant eff(vcf) file for filtering",
        required=True
    )
    parser.add_argument(
        "-o", "--output",
        dest="output_conf",
        help="output noSNP (vcf) file with known snps removed.",
        required=True
    )
    parser.add_argument(
        "-c", "--cov_margin",
        dest="cov_margin",
        help="minimum edit fraction",
        required=False,
        default=0.01,
        type=float
    )
    parser.add_argument(
        "-a", "--alpha",
        dest="alpha",
        help="alpha parameter",
        required=False,
        default=0,
        type=int
    )
    parser.add_argument(
        "-b", "--beta",
        dest="beta",
        help="beta parameter",
        required=False,
        default=0,
        type=int
    )
    parser.add_argument(
        "--keep-100-edited",
        dest="keep100",
        help="keep 100 percent edited sites (by default, these are tossed as SNPs)",
        required=False,
        default=False,
        action='store_true'
    )
    # Process arguments
    args = parser.parse_args()
    input_eff = args.input_eff
    outfile = args.output_conf
    cov_margin = args.cov_margin
    alpha = args.alpha
    beta = args.beta
    keep_all_edited = args.keep100

    #

    rank_edits(alpha, beta, cov_margin, keep_all_edited, input_eff, outfile)
    return 0



if __name__ == "__main__":
    sys.exit(main())
